File: benchmarks/java/Primes.java
// Prime sieve benchmark - tests loops and array operations
public class Primes {
    public static int countPrimes(int limit) {
        boolean[] sieve = new boolean[limit + 1];
        for (int i = 0; i <= limit; i++) sieve[i] = true;
        sieve[0] = sieve[1] = false;

        for (int i = 2; i * i <= limit; i++) {
            if (sieve[i]) {
                for (int j = i * i; j <= limit; j += i) {
                    sieve[j] = false;
                }
            }
        }

        int count = 0;
        for (int i = 2; i <= limit; i++) {
            if (sieve[i]) count++;
        }
        return count;
    }

    public static void main(String[] args) {
        int limit = 1000000;
        if (args.length > 0) limit = Integer.parseInt(args[0]);

        long start = System.nanoTime();
        int result = countPrimes(limit);
        long elapsed = (System.nanoTime() - start) / 1000000;

        System.out.println("Primes up to " + limit + ": " + result);
        System.out.println("Time: " + elapsed + " ms");
    }
}

File: benchmarks/zig/primes.zig
// Prime sieve benchmark - tests loops and array operations
const std = @import("std");

fn countPrimes(allocator: std.mem.Allocator, limit: usize) !usize {
    var sieve = try allocator.alloc(bool, limit + 1);
    defer allocator.free(sieve);

    for (sieve) |*s| s.* = true;
    sieve[0] = false;
    sieve[1] = false;

    var i: usize = 2;
    while (i * i <= limit) : (i += 1) {
        if (sieve[i]) {
            var j = i * i;
            while (j <= limit) : (j += i) {
                sieve[j] = false;
            }
        }
    }

    var count: usize = 0;
    for (sieve) |s| {
        if (s) count += 1;
    }
    return count;
}

pub fn main() !void {
    const stdout = std.io.getStdOut().writer();
    const args = try std.process.argsAlloc(std.heap.page_allocator);
    defer std.process.argsFree(std.heap.page_allocator, args);

    var limit: usize = 1000000;
    if (args.len > 1) {
        limit = try std.fmt.parseInt(usize, args[1], 10);
    }

    var timer = try std.time.Timer.start();
    const result = try countPrimes(std.heap.page_allocator, limit);
    const elapsed = timer.read() / std.time.ns_per_ms;

    try stdout.print("Primes up to {d}: {d}\n", .{ limit, result });
    try stdout.print("Time: {d} ms\n", .{elapsed});
}
